src/compiler/parser.ts

```diff
@@ -1537,6 +1537,138 @@ namespace ts {
+
+                case "type":
+                    parseErrorForExpectedName(scanner.getTokenText(), "=", Diagnostics.Type_alias_must_be_given_a_name, Diagnostics.Type_alias_name_cannot_be_0);
+                    return;
+            }
+
+            // The user alternately might have misspelled or forgotten to add a space after a common keyword.
+            const suggestion = getSpellingSuggestion(expressionText, viableKeywordSuggestions, n => n) || getSpaceSuggestion(expressionText);
+            if (suggestion) {
+                parseErrorAt(pos, expression.end, Diagnostics.Unknown_keyword_or_identifier_Did_you_mean_0, suggestion);
+                return;
+            }
+
+            // We know this is a slightly more precise case than a missing expected semicolon.
+            parseErrorAt(pos, expression.end, Diagnostics.Unexpected_keyword_or_identifier);
+        }
+
+        function parseErrorForExpectedName(name: string, normalToken: string, blankDiagnostic: DiagnosticMessage, nameDiagnostic: DiagnosticMessage) {
+            if (name === normalToken) {
+                parseErrorAtCurrentToken(blankDiagnostic);
+            }
+            else {
+                parseErrorAtCurrentToken(nameDiagnostic, name);
+            }
+        }
+
+        function getSpaceSuggestion(expressionText: string) {
+            for (const keyword of viableKeywordSuggestions) {
+                if (expressionText.length > keyword.length + 2 && startsWith(expressionText, keyword)) {
+                    return `${keyword} ${expressionText.slice(keyword.length)}`;
+                }
+            }
+
+            return undefined;
+        }
+
+        function parseSemicolonAfterPropertyName(name: PropertyName, type: TypeNode | undefined, initializer: Expression | undefined) {
+            switch (scanner.getTokenText()) {
+                case "@":
+                    parseErrorAtCurrentToken(Diagnostics.Decorators_must_precede_all_other_keywords_for_property_declarations);
+                    return;
+
+                case "(":
+                    parseErrorAtCurrentToken(Diagnostics.Function_call_is_not_a_type_annotation);
+                    nextToken();
+                    return;
+            }
+
+            if (type && !canParseSemicolon()) {
+                if (initializer) {
+                    parseErrorAtCurrentToken(Diagnostics._0_expected, tokenToString(SyntaxKind.SemicolonToken));
+                }
+                else {
+                    parseErrorAtCurrentToken(Diagnostics.Missing_before_default_property_value);
+                }
+                return;
+            }
+
+            return parseSemicolonAfter(SyntaxKind.PropertyDeclaration, name, initializer);
+        }
+
+        function getExpressionText(expression: Expression | PropertyName) {
+            return expression && ts.isIdentifier(expression)
+                ? expression.escapedText.toString()
+                : undefined;
+        }

```

Comment:
See if

* `entityNameToString`
* `getPropertyNameForPropertyNameNode`
* `tryGetPropertyAccessOrIdentifierToString`
* `declarationNameToString`

do the job, since there's already 4 functions like this

Reply:
* `entityNameToString` does not accept `Expression`s nor `PropertyName`s
* `getPropertyNameForPropertyNameNode` does not accept `Expression`s
* `tryGetPropertyAccessOrIdentifierToString` does not accept `PropertyName`s
* `declarationNameToString` does not accept `Expression`s

I _could_ go with `getTextOfNode`, but that feels a little... hacky? Is my instinct right there?

Reply:
I think you want `idText(...)` instead of `expression.escapedText.toString()` to do the unescaping.